<a href="https://colab.research.google.com/github/Sharonsellnorbert/DSA_B13_casestudy/blob/main/case_study_Data_Acquisition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [23]:
#Task 1

In [24]:
import pandas as pd
import numpy as np

In [25]:
po=pd.read_csv('/content/population.csv')

In [26]:
po.head(5)

,country,population
0,United States,331002651
1,China,1411778724
2,India,1380004385
3,Japan,125836021
4,Germany,83240525


In [27]:
po.shape

(25, 2)

In [28]:
po.dtypes

,0
country,object
population,int64


In [29]:
# TASK 2

In [30]:
gd=pd.read_excel('/content/gdp.xlsx')

In [31]:
gd.columns

Index(['Country', 'GDP'], dtype='object')

In [32]:
gd.dtypes

,0
Country,object
GDP,int64


In [33]:
# TASK 3

In [34]:
iu=pd.read_json('/content/internet_users.json')

In [35]:
df= pd.DataFrame(iu)

In [36]:
df

,country,internet_users
0,United States,312000000
1,China,1080000000
2,India,850000000
3,Japan,118000000
4,Germany,79000000
5,United Kingdom,66000000
6,France,62000000
7,Italy,56000000
8,Canada,36000000
9,Australia,24000000


In [37]:
# TASK 4

In [38]:
x=pd.read_xml('/content/literacy_rate.xml')

In [39]:
dfx=pd.DataFrame(x)

In [40]:
dfx

,name,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0


In [41]:
# TASK 5

In [42]:
# Explore Population dataset
print("POPULATION")
print("Rows and Columns:", po.shape)
print("Column Names:", po.columns)
print("Data Types:")
print(po.dtypes)
print("Missing Values:")
print(po.isnull().sum())
print("Duplicate Records:", po.duplicated().sum())


# Explore GDP dataset
print("\nGDP")
print("Rows and Columns:", gd.shape)
print("Column Names:", gd.columns)
print("Data Types:")
print(gd.dtypes)
print("Missing Values:")
print(gd.isnull().sum())
print("Duplicate Records:", gd.duplicated().sum())


# Explore Internet Users dataset
print("\nINTERNET USERS")
print("Rows and Columns:", df.shape)
print("Column Names:", df.columns)
print("Data Types:")
print(df.dtypes)
print("Missing Values:")
print(df.isnull().sum())
print("Duplicate Records:", df.duplicated().sum())


# Explore Literacy Rate dataset
print("\nLITERACY RATE")
print("Rows and Columns:", dfx.shape)
print("Column Names:", dfx.columns)
print("Data Types:")
print(dfx.dtypes)
print("Missing Values:")
print(dfx.isnull().sum())
print("Duplicate Records:", dfx.duplicated().sum())

POPULATION
Rows and Columns: (25, 2)
Column Names: Index(['country', 'population'], dtype='object')
Data Types:
country       object
population     int64
dtype: object
Missing Values:
country       0
population    0
dtype: int64
Duplicate Records: 0

GDP
Rows and Columns: (25, 2)
Column Names: Index(['Country', 'GDP'], dtype='object')
Data Types:
Country    object
GDP         int64
dtype: object
Missing Values:
Country    0
GDP        0
dtype: int64
Duplicate Records: 0

INTERNET USERS
Rows and Columns: (25, 2)
Column Names: Index(['country', 'internet_users'], dtype='object')
Data Types:
country           object
internet_users     int64
dtype: object
Missing Values:
country           0
internet_users    0
dtype: int64
Duplicate Records: 0

LITERACY RATE
Rows and Columns: (25, 2)
Column Names: Index(['name', 'literacy_rate'], dtype='object')
Data Types:
name              object
literacy_rate    float64
dtype: object
Missing Values:
name             0
literacy_rate    0
dtype: int64
Dup

In [43]:
po.rename(columns={'country': 'country'})

gd.rename(columns={'Country': 'country'})

df.rename(columns={'country': 'country'})

dfx.rename(columns={'name': 'country'})

,country,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0


In [44]:
print(po.dtypes)
print(gd.dtypes)
print(df.dtypes)
print(dfx.dtypes)

country       object
population     int64
dtype: object
Country    object
GDP         int64
dtype: object
country           object
internet_users     int64
dtype: object
name              object
literacy_rate    float64
dtype: object


In [50]:
# Rename columns so that all datasets have a common key
gd.rename(columns={'Country': 'country'}, inplace=True)
dfx.rename(columns={'name': 'country'}, inplace=True)

# Check column names
print(po.columns)
print(gd.columns)
print(df.columns)
print(dfx.columns)

# Merge Population and GDP
merged = pd.merge(po, gd, on='country')

# Merge with Internet Users
merged = pd.merge(merged, df, on='country')

# Merge with Literacy Rate
merged = pd.merge(merged, dfx, on='country')

# Display merged dataset
display(merged)


Index(['country', 'population'], dtype='object')
Index(['country', 'GDP'], dtype='object')
Index(['country', 'internet_users'], dtype='object')
Index(['country', 'literacy_rate'], dtype='object')


,country,population,GDP,internet_users,literacy_rate
0,United States,331002651,30500000,312000000,99.0
1,China,1411778724,19230000,1080000000,97.2
2,India,1380004385,4187000,850000000,77.7
3,Japan,125836021,4180000,118000000,99.0
4,Germany,83240525,4744000,79000000,99.0
5,United Kingdom,67886011,3839000,66000000,99.0
6,France,65273511,3211000,62000000,99.0
7,Italy,60461826,2420000,56000000,99.0
8,Canada,38005238,2250000,36000000,99.0
9,Australia,25687041,1880000,24000000,99.0


In [51]:
merged['Internet_penetration']=(merged['internet_users']/merged['population']*100).round(2)
merged['GDP Per Capita']=(merged['GDP']/merged['population']).round(2)
merged

,country,population,GDP,internet_users,literacy_rate,Internet_penetration,GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
2,India,1380004385,4187000,850000000,77.7,61.59,0.00
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06
5,United Kingdom,67886011,3839000,66000000,99.0,97.22,0.06
6,France,65273511,3211000,62000000,99.0,94.98,0.05
7,Italy,60461826,2420000,56000000,99.0,92.62,0.04
8,Canada,38005238,2250000,36000000,99.0,94.72,0.06
9,Australia,25687041,1880000,24000000,99.0,93.43,0.07


In [52]:
import sqlite3

conn = sqlite3.connect('/content/country_statistics.db')

merged.to_sql(
    'country_statistics',
    conn,
    if_exists='replace',
    index=False
)
print("Data successfully loaded into SQLite database.")

Data successfully loaded into SQLite database.


In [53]:
# Retrieve first 5 records
query = "SELECT * FROM country_statistics LIMIT 5;"

result = pd.read_sql_query(query, conn)

result

,country,population,GDP,internet_users,literacy_rate,Internet_penetration,GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
2,India,1380004385,4187000,850000000,77.7,61.59,0.00
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06


In [54]:
# 1. Top 5 countries by GDP
print(merged.sort_values('GDP', ascending=False).head(5))
print('---------------------------------------------------------')
# 2. Top 5 countries by population
print(merged.sort_values('population', ascending=False).head(5))
print('---------------------------------------------------------')
# 3. Average literacy rate
print(merged['literacy_rate'].mean())
print('---------------------------------------------------------')
# 4. Literacy rate greater than 90%
print(merged[merged['literacy_rate'] > 90])
print('---------------------------------------------------------')
# 5. Internet penetration greater than 70%
print(merged[merged['Internet_penetration'] > 70])
print('---------------------------------------------------------')
# 6. Top 5 GDP per capita
print(merged.sort_values('GDP Per Capita', ascending=False).head(5))
print('---------------------------------------------------------')
# 7. Total population
print(merged['population'].sum())
print('---------------------------------------------------------')
# 8. Average internet penetration
print(merged['Internet_penetration'].mean())

print('---------------------------------------------------------')
# 9. Country with highest internet users
print(merged.loc[
    merged['internet_users'].idxmax(),
    ['country', 'internet_users']
])
print('---------------------------------------------------------')
# 10. Correlation
print(merged['literacy_rate'].corr(
    merged['Internet_penetration']
))

         country  population       GDP  internet_users  literacy_rate  \
0  United States   331002651  30500000       312000000           99.0   
1          China  1411778724  19230000      1080000000           97.2   
4        Germany    83240525   4744000        79000000           99.0   
2          India  1380004385   4187000       850000000           77.7   
3          Japan   125836021   4180000       118000000           99.0   

   Internet_penetration  GDP Per Capita  
0                 94.26            0.09  
1                 76.50            0.01  
4                 94.91            0.06  
2                 61.59            0.00  
3                 93.77            0.03  
---------------------------------------------------------
          country  population       GDP  internet_users  literacy_rate  \
1           China  1411778724  19230000      1080000000           97.2   
2           India  1380004385   4187000       850000000           77.7   
0   United States   331002651

### SQLite Database Analysis (Questions 11 to 16)

In [55]:
import pandas as pd
import sqlite3

# Open database connection
conn = sqlite3.connect('/content/country_statistics.db')

# Ensure the table is created/populated in the DB from the in-memory 'merged' dataframe
merged.to_sql('country_statistics', conn, if_exists='replace', index=False)

# 11. Top 5 countries by GDP
q11 = """
SELECT country, GDP
FROM country_statistics
ORDER BY GDP DESC
LIMIT 5;
"""
print("11. Top 5 countries by GDP:")
display(pd.read_sql_query(q11, conn))

11. Top 5 countries by GDP:


,country,GDP
0,United States,30500000
1,China,19230000
2,Germany,4744000
3,India,4187000
4,Japan,4180000


In [56]:

q12 = """
SELECT AVG(literacy_rate) AS average_literacy_rate
FROM country_statistics;
"""
print("12. Average literacy rate:")
display(pd.read_sql_query(q12, conn))

12. Average literacy rate:


,average_literacy_rate
0,97.248


In [57]:

q13 = """
SELECT country, literacy_rate
FROM country_statistics
WHERE literacy_rate > 90.0;
"""
print("13. Countries with literacy rate above 90%:")
display(pd.read_sql_query(q13, conn))

13. Countries with literacy rate above 90%:


,country,literacy_rate
0,United States,99.0
1,China,97.2
2,Japan,99.0
3,Germany,99.0
4,United Kingdom,99.0
5,France,99.0
6,Italy,99.0
7,Canada,99.0
8,Australia,99.0
9,Brazil,93.2


In [58]:

q14 = """
SELECT country, Internet_penetration
FROM country_statistics
ORDER BY Internet_penetration DESC
LIMIT 5;
"""
print("14. Top 5 countries with highest internet penetration:")
display(pd.read_sql_query(q14, conn))

14. Top 5 countries with highest internet penetration:


,country,Internet_penetration
0,Switzerland,98.21
1,Norway,97.76
2,Saudi Arabia,97.66
3,Sweden,97.55
4,Netherlands,97.47


In [59]:

q15 = """
SELECT SUM(population) AS total_population
FROM country_statistics;
"""
print("15. Total population:")
display(pd.read_sql_query(q15, conn))

15. Total population:


,total_population
0,4720039725


In [60]:

q16 = """
SELECT country, "GDP Per Capita"
FROM country_statistics
ORDER BY "GDP Per Capita" DESC
LIMIT 1;
"""
print("16. Country with the highest GDP per capita:")
display(pd.read_sql_query(q16, conn))

16. Country with the highest GDP per capita:


,country,GDP Per Capita
0,Switzerland,0.11


### Export Dataset and Sort (Questions 17 & 18)

In [61]:
import pandas as pd
import sqlite3


conn = sqlite3.connect('/content/country_statistics.db')
merged = pd.read_sql_query("SELECT * FROM country_statistics", conn)


merged.to_csv('/content/final_merged_countries.csv', index=False)
print("Dataset exported successfully to /content/final_merged_countries.csv")

Dataset exported successfully to /content/final_merged_countries.csv


In [62]:

top_10_gdp_capita = merged.sort_values(by='GDP Per Capita', ascending=False).head(10)
display(top_10_gdp_capita)

,country,population,GDP,internet_users,literacy_rate,Internet_penetration,GDP Per Capita
21,Switzerland,8654622,940000,8500000,99.0,98.21,0.11
23,Norway,5421241,580000,5300000,100.0,97.76,0.11
24,Singapore,5850342,530000,5600000,97.5,95.72,0.09
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
9,Australia,25687041,1880000,24000000,99.0,93.43,0.07
20,Netherlands,17441139,1150000,17000000,99.0,97.47,0.07
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06
8,Canada,38005238,2250000,36000000,99.0,94.72,0.06
5,United Kingdom,67886011,3839000,66000000,99.0,97.22,0.06
22,Sweden,10353442,650000,10100000,99.0,97.55,0.06


Summary of the ETL and Analysis Process
Diverse Data Sources:
We extracted data from four distinct file formats: a CSV file for population records (population.csv), an Excel spreadsheet for GDP statistics (gdp.xlsx), a JSON file for internet user counts (internet_users.json), and an XML document for literacy rates (literacy_rate.xml).

Targeted Data Cleaning:
Key column names were standardized (e.g., renaming 'Country' to 'country' and 'name' to 'country') to establish a uniform key for merging, and we validated that there were no duplicate entries or missing values across the tables.

Comprehensive Transformations:
All four datasets were merged on the 'country' column into a unified DataFrame. We then calculated two custom metrics: Internet Penetration (%) (internet users divided by population) and GDP Per Capita (GDP divided by population).

Robust Database Loading:
We established a connection to a local SQLite database and created a table named country_statistics using Python's sqlite3 and pandas libraries, successfully loading all merged data for SQL-based querying. Key Analytical Insights:

GDP & Population:
The United States has the highest total GDP, while China has the largest population in our dataset.

Development Indicators:
Switzerland boasts the highest GDP per capita, and Norway has the highest internet penetration (97.76%) coupled with a 100% literacy rate.

Correlation:
We found a strong positive correlation (~0.77) between a country's literacy rate and its internet penetration percentage.

Nandagopan_L
nexus_9317
